### PR TITLE
[REG-1273] Harden network connectivity and re-connects

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
Not Perfect, but should reconnect remote bots automatically in most scenarios.

- Thread safety / concurrency fixes around network connection management
  - Resolves an issue where sometimes we wouldn't re-connect as necessary due to a concurrency issue setting/reading a variable
- Changes to better detect most socket state failures before read and write
- Added socket read/write timeouts to make failures detect properly
- Fixed socket read code to address issues handling transient errors / empty buffer during read attempts

### CODE REVIEW

File: Runtime/Scripts/RGBotSpawnManager.cs [hunk 1]
```diff
@@ -280,7 +280,15 @@ protected internal void CallSeatBot(BotInformation botToSpawn)
...
                 // First, allow the developer to configure any frontend/backend for a player about to spawn,
                 // such as character selection.
-                SeatBot(botToSpawn);
+                try
+                {
+                    SeatBot(botToSpawn);
+                }
+                catch (Exception ex)
+                {
+                    RGDebug.LogError($"Error seating bot - {ex}");
```
Comment (vontell) [inline Runtime/Scripts/RGBotSpawnManager.cs:289]: Should we also just return here or raise an exception? Not sure if we want to continue getting the bot ready if the seating failed.

**Review by vontell — CHANGES_REQUESTED** [review]
Just one question from me in the code. Also, should we add a UAT, or I'm assuming this will be covered by existing UAT as well?

File: Runtime/Scripts/RGBotSpawnManager.cs (hunk 1 above)
Comment (RG-nAmKcAz) [reply Runtime/Scripts/RGBotSpawnManager.cs]: removed

**Review by vontell — APPROVED** [review]

File: Runtime/Scripts/Types/RGClientConnection_Remote.cs [hunk 2]
```diff
@@ -114,113 +143,116 @@ public override bool Connected()
...
+            {
+                lock (this)
+                {
+                    if (!_connecting && !Connected())
+                    {
+                        _connecting = true;
+                        shouldIConnect = true;
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply Runtime/Scripts/Types/RGClientConnection_Remote.cs]: I've never used Lazy for a primitive value before. In c# primitives are atomic for a single operation.  There are still limits to cross thread visibility, especially in multi processor systems, but as long as you keep those conditions in mind you can avoid the overhead of Lazy's lock for everything pattern.

I avoided using lazy here as `ExecutionAndPublication` always locks when checking its value and the intent of the double checked locking pattern is to avoid locking when not necessary.  As I understand it, it would be roughly equivalent to Lazy's `ExecutionAndPublication` from a locking perspective to just lock and then check instead of using double checked locking; not exactly the same cpu cost, but logically similar.

In this particular case you could argue that double checked locking is only truly safe for preventing 2 simultaneous connection attempts, but may possibly miss `_connecting` transitioning to false as set by another thread and not connect in this method call when it technically could.  This statement is true... however, the system will still see on the next pass (RGBotServerListener.Update) that we aren't connected and will attempt the connection again.  This outer code already exists to detect random disconnect scenarios and is not specific to interactions with this locking code.

Long story short, I updated the code to use a Semaphore here instead of a boolean.  It doesn't change anything from a functionality perspective as described in the prior sentences, but it may be a bit more clear for others reading the code.

File: Runtime/Scripts/Types/RGClientConnection_Remote.cs [hunk 3]
```diff
@@ -61,13 +66,40 @@ private bool SendSocketMessage(string type, string data)
...
+                {
+                    var vt = _client.GetStream().WriteAsync(finalBuffer);
+                    vt.ConfigureAwait(false).GetAwaiter().OnCompleted(() =>
+                    {
+                        // see if the socket is no longer writeable
+                        if (!vt.IsCompletedSuccessfully)
+                        {
+                            HandleDeadWriteState();
+                        }
+                    });
```
Comment (addisonbgross) [inline Runtime/Scripts/Types/RGClientConnection_Remote.cs:80]: is there a need to implement a retry policy if this write fails?

Comment (RG-nAmKcAz) [reply Runtime/Scripts/Types/RGClientConnection_Remote.cs:80]: updates are not guaranteed to get there.  If the connection is down, all the states between 'now' and 'future now' are not forwarded to the client as they should have no need to process them anyway

in fact.. on the client side if it is processing slowly and we get multiple new states before we finish processing old states, we also explicitly trash the states in between

**Review by rcornwall-reg — APPROVED** [review]
